Improve Error message of invalid user build option

GithubMaddin opened the pull request:
The topic of better error messages for user build options was brought up in issue #6797. The commit 17c8193 partially covered the issue by adding details to the error message for one specific user build option that was mentioned in the issue.
This pull request however applies the same treatment to all types of user build options with the aim to enhance the general usefulness of these error messages. This closes the issue #6797.
The error messages now print out the description of each option in addition to the value to let the user easily know which option has an invalid value.
Additionally due to the changes in the error messages, the test cases previously added by commit 17c8193 have been adjusted accordingly.

mensinda commented:
Instead of logging the description, I would say logging the name of the option would be better (since the name is what the user actually interacts with most of the time). Especially, since the descriptions can be rather lengthy.

I am aware that the options currently don't store their own name, so this would involve a bit more refactoring. Feel free to ask if you have any questions.

BlogThe commented:
Hi Daniel, 
thanks for the feedback. We (GithubMaddin and me worked together on this) agree with you that logging the name is the better option. However we were hesitant to implement this, since this would require extensive rewriting of existing functions and constructors in the coredata.py and other files. We have now decided to follow your advice and started rewriting the constructor for UserOption and the specific subtypes (UserStringOption, UserIntegerOption, etc.). This has worked for the most part and we can now display the name in the error messages for build options provided inside the meson_options.txt. Additionally we tried to find more instances of a user build option being created. It seems that there are several several places in the code where one of the user option types is created. When testing the -D command line option to set the user build options, we encountered an error.

```
[...]

  File "/home/martin/meson-fork/mesonbuild/coredata.py", line 89, in set_value
    self.value = self.validate_value(newvalue)
  File "/home/martin/meson-fork/mesonbuild/coredata.py", line 142, in validate_value
    raise MesonException(f'Value "{value}" for {self.opt_type} option "{self.opt_name}" is more than maximum value {self.max_value}.')
AttributeError: 'UserIntegerOption' object has no attribute 'opt_type'
```

This error was triggered when using the following command:

```  ../meson-fork/meson.py -Dinteger_opt=15 --reconfigure builddir/ ```

We tested the same scenario (an integer option exceeding it's max value) with the meson_options.txt and didn't encounter this error.  So it seems like there are some places in the code where an instance of an option is created without using the constructor. At the moment we are a bit unsure how to best identify these kind of occurences. Do you maybe have some advice in that regard?

mensinda commented:
Unfortunately, there is not much that I can help you with without any source code and also debugging the issue myself. The only advice I could give is to search for `User[a-zA-Z]+Option` in VS-Code (but I guess that you are already doing that)...

Refactoring-wise, I would advise you to pass `OptionKey`s for the name to the constructor, instead of strings. Doing so makes writing factory functions easier to avoid code/name duplications.

mensinda commented:
Another point is that we are serializing the entire `CoreData` class to disk with [`pickle`](https://docs.python.org/3/library/pickle.html). So, there can be hard to debug issues if you change something in your code and have an already existing build directory. You could also try checking if everything is correctly unpickled just in case.

Not sure if this helps, but this is all I can currently think of.

BlogThe commented:
We squashed the previous commits to clean up the commit history a bit. We are still pretty new to contributing to open-source projects and didn't think about the messy commits. However no committed code changes since the last one, only cleanup so far.

GithubMaddin commented:
Hi Daniel,
with the previous commit we have added the name and the type of the user option to the UserOption Classes. They are now available as attribute opt_name and opt_type. 
Additional we spent a lot ot time to change all corresponding calls of the constructors. 

So far we tested the new setup with meson_options.txt and it worked as expected.

When testing the same setup for an UserIntegerOption with the -D command. we received an error message in cases that the option name contained an underscore.

To give you some more details:

For testing purposes we created a test project including a meson_options.txt
We added two Integer options:
```
# integer option
option('abcde', type : 'integer', min : 2, max : 5, value: 3) # correct
option('integer_opt', type : 'integer', min : 2, max : 5, value: 3) # correct
```
When testing (option with underscore): 
```
../meson-fork3/meson.py -Dinteger_opt=1 --reconfigure builddir/
```

we receive this error message:

```
The Meson build system
Version: 0.58.999
Source dir: /home/martin/test_meson
Build dir: /home/martin/test_meson/builddir
Build type: native build
Traceback (most recent call last):
  File "/home/martin/meson-fork3/mesonbuild/mesonmain.py", line 223, in run
    return options.run_func(options)
  File "/home/martin/meson-fork3/mesonbuild/msetup.py", line 281, in run
    app.generate()
  File "/home/martin/meson-fork3/mesonbuild/msetup.py", line 184, in generate
    self._generate(env)
  File "/home/martin/meson-fork3/mesonbuild/msetup.py", line 201, in _generate
    intr = interpreter.Interpreter(b)
  File "/home/martin/meson-fork3/mesonbuild/interpreter/interpreter.py", line 362, in __init__
    self.parse_project()
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/interpreterbase.py", line 117, in parse_project
    self.evaluate_codeblock(self.ast, end=1)
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/interpreterbase.py", line 163, in evaluate_statement
    return self.function_call(cur)
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/interpreterbase.py", line 549, in function_call
    res = func(node, func_args, kwargs)
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/decorators.py", line 64, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/martin/meson-fork3/mesonbuild/interpreterbase/decorators.py", line 102, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/martin/meson-fork3/mesonbuild/interpreter/interpreter.py", line 1117, in func_project
    self.coredata.set_default_options(default_options, self.subproject, self.environment)
  File "/home/martin/meson-fork3/mesonbuild/coredata.py", line 833, in set_default_options
    self.set_options(options, subproject=subproject)
  File "/home/martin/meson-fork3/mesonbuild/coredata.py", line 790, in set_options
    self.set_option(k, v)
  File "/home/martin/meson-fork3/mesonbuild/coredata.py", line 641, in set_option
    self.options[key].set_value(value)
  File "/home/martin/meson-fork3/mesonbuild/coredata.py", line 91, in set_value
    self.value = self.validate_value(newvalue)
  File "/home/martin/meson-fork3/mesonbuild/coredata.py", line 142, in validate_value
    raise MesonException(f'Value "{value}" for {self.opt_type} option "{self.opt_name}" is less than minimum value {self.min_value}.')
AttributeError: 'UserIntegerOption' object has no attribute 'opt_type'
```
When testing the same command with the other option (without underscore): 
```
../meson-fork3/meson.py -Dabcde=1 --reconfigure builddir/
```
we received the expected result:
```
The Meson build system
Version: 0.58.999
Source dir: /home/martin/test_meson
Build dir: /home/martin/test_meson/builddir
Build type: native build

meson.build:1:0: ERROR: Value "1" for integer option "abcde" is less than minimum value 2.

```


We tried several runs to identify the source of the issue. 
So far it seems that the UserOption is not created properly.

As you can see from the error message the error, that the value is out of bound was detected properly, so there seems to be access to the bounderies specified in meson_options.txt. However, the UserIntegerOption-object does not seem to have the proper attributes. 

We assume that is it related with the BuiltinOption() generic class that creates user options. (mesonbuild/coredata.py: 1088)
We are not that experienced with python that we were definetly able to identify how this class actually create an user option.

We would be glad if someone could support us in finding the cause of the issue. 
So far we don't see any option to fix it ourself.

mensinda commented:
I am unable to reproduce this error with your current PR. I am getting another error instead:

```
  File "/home/daniel/projects/meson/mesonbuild/coredata.py", line 1111, in init_option
    return self.opt_type(self.description, **keywords)
TypeError: __init__() missing 1 required positional argument: 'opt_name'
```

`BuiltinOption` creates the actual option object [here](https://github.com/mesonbuild/meson/blob/master/mesonbuild/coredata.py#L1102). This is because `BuiltinOption` is often called like this `BuiltinOption(UserStringOption, ...)`. Thus, the `self.opt_type` stores the *class* type and not an object. Then calling `self.opt_type()` will be equivalent to `UserStringOption()`.

As a general tip: Use `print()` debugging (maybe in combination with `traceback.print_stack()` to get the call stack) when you are not sure when and what code is executed.

Also, as a general git tip: Use a separate branch (*not* master) in your fork for development. This way, you are still able to pull the current changes from the main repo master branch. Also, never merge master into your new branch. Always use `git rebase master` instead.

GithubMaddin commented:
Hi Daniel, thank you for the tips and the comment regarding the BuiltinOption. 
Unfortunately we are quite busy at the moment. We give it another try in the upcoming days/weeks.
Hope we will find a solution for our and your error.

GithubMaddin commented:
Hi Daniel,
Your latest comment regarding calling the constructor for buildoptions was the solution for the issue.
We tested it again and the errors that occurred did not appear again using the same setup.
Most likely the error message that was shown to you should also be fixed with this last minor change of our code. 
We were not able to replicate your error but based on your error description we are certain the error is solved now. 
We tried to get our PR up-to-date to the status of the master using your recommendation with git rebase. Locally it worked. Our testing was successful, unfortunately the checks of the pull-requests fail.
It seems like the test don’t like the way we force pushed the master branch:
e.g.: “This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.” (https://github.com/mesonbuild/meson/commit/4c1f1670a3880b8fce8fb811f3b57b6d9eb329e5#annotation_1691252094)
But we don’t know how to fix it.
It really is a pity because is seem like only two lines were missing (the lines in the build-options you pointed out). Unfortunately, there have been a few pull requests in the last weeks that caused changes in files that also have been affected with our changes. 
Due to the fact that our exam phase will start in a couple of days we are not sure if we will be able to add further changes to our pull request. 
We don’t know how we could fix the issue shown by the auto-git-test.
Is it possible that you can have a look at it and fix it, we will not have the capacity and if we include the change in a couple of month it will become even more complicated to include it in this ongoing project.
We would be very sad if all the work that we already spent to include the changes would be lost hours. So we would be grateful if somebody could have a look at it and might fix the last issue.

Best regards,

Dominik and Martin

eli-schwartz commented:
Hi! I wanted to take another look at this PR, so I have gone ahead and rebased it, resolving a bunch of conflicts. I'm not completely positive I made a perfect rebase, but I think I did a decent job. Just in case, I pushed the old version of the PR to the branch `pr8895-old`.

There are definitely intervening issues that popped up, CI is failing with TypeErrors due to newly added options. Hopefully this rebase helps you get back on the road to fully implementing the improvement!

(Also, we're currently in release freeze so actual churn to the repository will be minimal for a while. If you work on this now, you won't need to wrangle other peoples' changes at the same time.)

eli-schwartz commented on mesonbuild/coredata.py:
```diff
@@ -1179,41 +1188,41 @@ def add_to_argparse(self, name: str, parser: argparse.ArgumentParser, help_suffi
+    (OptionKey('force_fallback_for'), BuiltinOption(UserArrayOption, 'Force fallback for those subprojects', [], 'force_fallback_for')),
 
     # Python module
     (OptionKey('platlibdir', module='python'),
```
These new options need to get opt_name attributes too. Should be easy to fix.

eli-schwartz commented:
Ping?
